processing/src/main/java/io/druid/query/groupby/epinephelinae/LimitedBufferGrouper.java

```diff
@@ -0,0 +1,536 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.groupby.epinephelinae;
+
+import com.google.common.base.Supplier;
+import com.google.common.collect.Iterators;
+import com.google.common.primitives.Ints;
+import io.druid.java.util.common.IAE;
+import io.druid.java.util.common.ISE;
+import io.druid.query.aggregation.AggregatorFactory;
+import io.druid.segment.ColumnSelectorFactory;
+
+import java.nio.ByteBuffer;
+import java.util.AbstractList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+public class LimitedBufferGrouper<KeyType> extends AbstractBufferGrouper<KeyType>
+{
+  private static final int MIN_INITIAL_BUCKETS = 4;
+  private static final int DEFAULT_INITIAL_BUCKETS = 1024;
+  private static final float DEFAULT_MAX_LOAD_FACTOR = 0.7f;
+
+  private final AggregatorFactory[] aggregatorFactories;
+
+  // Limit to apply to results.
+  // If limit > 0, track hash table entries in a binary heap with size of limit.
+  // If -1, no limit is applied, hash table entry offsets are tracked with an unordered list with no limit.
+  private int limit;
```

Comment:
This was an outdated comment from an older revision, limit will never be <= 0 now, I deleted the comment